lender.hpp

```diff
@@ -196,10 +202,41 @@ class library : public recycle_memory<T> {
     auto operate() -> buffer_ptr<T>;
 };
 
+template <typename T>
+class mailbox : public recycle_memory<T> {
+    friend class mail_test;
+
+    private:
+        int max_read;
+        // TODO: does map_value need to be a template or not?
+        // template <typename T>
+        struct map_value {
+            std::mutex val_lock;

```

Comment:
It's possible that you might be able to remove `val_lock` (and `val_cd`) from this class. It appears that the `value` and `count` members are only ever accessed when the `mailbox::box_lock` has been acquired first; which would seem to provide adequate protection against race conditions. Of course, I could be missing something.